Modules/_testcapi/long.c

```diff
@@ -141,13 +141,155 @@ pylong_aspid(PyObject *module, PyObject *arg)
 }
 
 
+static PyObject *
+layout_to_dict(const PyLongLayout *layout)
+{
+    PyObject *dict = PyDict_New();
+    if (dict == NULL) {
+        goto error;
+    }
+
+#define SET_DICT(KEY, EXPR) \
+    do { \
+        PyObject *value = (EXPR); \
+        if (value == NULL) { \
+            goto error; \
+        } \
+        int res = PyDict_SetItemString(dict, KEY, value); \
+        Py_DECREF(value); \
+        if (res < 0) { \
+            goto error; \
+        } \
+    } while (0)
+
+    SET_DICT("bits_per_digit", PyLong_FromUnsignedLong(layout->bits_per_digit));
+    SET_DICT("digit_size", PyLong_FromUnsignedLong(layout->digit_size));
+    SET_DICT("digits_order", PyLong_FromLong(layout->digits_order));
+    SET_DICT("digit_endianness", PyLong_FromLong(layout->digit_endianness));
+#undef SET_DICT
+
+    return dict;
+
+error:
+    Py_XDECREF(dict);
+    return NULL;
+}
+
+
+static PyObject *
+pylong_export(PyObject *module, PyObject *obj)
+{
+    PyLongExport export_long;
+    if (PyLong_Export(obj, &export_long) < 0) {
+        return NULL;
+    }
+
+    if (export_long.digits == NULL) {
+        return PyLong_FromInt64(export_long.value);
+        // PyLong_FreeExport() is not needed in this case
+    }
+
+    assert(PyLong_GetNativeLayout()->digit_size == sizeof(digit));
+    const digit *export_long_digits = export_long.digits;
+
+    PyObject *digits = PyList_New(0);
+    if (digits == NULL) {
+        goto error;
+    }
+    for (Py_ssize_t i = 0; i < export_long.ndigits; i++) {
+        PyObject *item = PyLong_FromUnsignedLong(export_long_digits[i]);
+        if (item == NULL) {
+            goto error;
+        }
+
+        if (PyList_Append(digits, item) < 0) {
+            Py_DECREF(item);
+            goto error;
+        }
+        Py_DECREF(item);
+    }
+
+    PyObject *res = Py_BuildValue("(iN)", export_long.negative, digits);
+
+    PyLong_FreeExport(&export_long);
+    assert(export_long._reserved == 0);
+
+    return res;
+
+error:
+    Py_XDECREF(digits);
+    PyLong_FreeExport(&export_long);
+    return NULL;
+}
+
+
+static PyObject *
+pylongwriter_create(PyObject *module, PyObject *args)
+{
+    int negative;
+    PyObject *list;
+    if (!PyArg_ParseTuple(args, "iO!", &negative, &PyList_Type, &list)) {
```

Comment:
Note that this does not allow to test negative `ndigits` and `digits == NULL`. Add a TODO if you do not want to test them right now.